# 02_Central_Tendency

### What are Measures of Central Tendency?

Measures of Central Tendency are statistical methods used to find the **central or typical value** of a dataset. The main measures are **Mean, Median, and Mode**. They help summarize data and are widely used in **statistics, data analysis, and AI/ML** to understand datasets and make better decisions.


## 01. Mean
**Definition**: Mean is the average of a set of numbers. It is calculated by adding all the values and dividing by the number of values.

**Formula**: Mean (x̄) = (Σx) / n

where Σx = sum of all values, n = number of values

**Real-World Example**: A cricket player's batting average is the mean of his runs scored across matches.

**Business Example**: A company calculates the average monthly revenue over the last 6 months to understand overall business trends.

**AI/ML Example**: Before training a model, the mean of a feature (like "Age") is often used to fill missing values.

In [1]:
marks = [80, 75, 90, 95, 88]

mean = sum(marks) / len(marks)
print("Mean:", mean)

Mean: 85.6


### Interpretation:
The average score of the students is 85.6, which gives a single summary value to represent the overall performance of the class.

## 02. Median
**Definition**: Median is the middle value of a dataset when it is arranged in ascending or descending order. If there are two middle values, the median is their average.

**Formula**: If n is odd: Median = value at position (n+1)/2

If n is even: Median = average of values at positions n/2 and (n/2)+1

**Real-World Example**: The median house price in a city is often reported instead of the mean, because a few very expensive houses can distort the mean.

**Business Example**: A company reports the median salary of its employees to give a fairer picture of typical pay, since a few top executives' salaries could skew the mean.

**AI/ML Example**: Median is used to fill missing values in a feature when the data has outliers, since median is not affected by extreme values.

In [2]:
import statistics

marks = [80, 75, 90, 95, 88]

median = statistics.median(marks)
print("Median:", median)

Median: 88


### Interpretation:
When the 5 marks are sorted (75, 80, 88, 90, 95), the middle value is 88. This tells us that half the students scored below 88 and half scored above it.

## 03. Mode
**Definition**: Mode is the value that occurs most frequently in a dataset. A dataset can have one mode, more than one mode, or no mode at all.

**Formula**: No mathematical formula — mode is found by counting the frequency of each value and selecting the most frequent one.

**Real-World Example**: The most common shoe size sold in a store is the mode of shoe sizes.

**Business Example**: A retail store identifies the most frequently purchased product (mode) to decide what to stock more of.

**AI/ML Example**: Mode is used to fill missing values in categorical features (e.g., filling missing "City" values with the most common city).

In [3]:
import statistics

sizes = [7, 8, 8, 9, 8, 7, 10]

mode = statistics.mode(sizes)
print("Mode:", mode)

Mode: 8


### Interpretation:
Size 8 appears most often (3 times) among the sold shoe sizes, so the store should stock more of size 8 compared to other sizes.

## 04. Weighted Mean
**Definition**: Weighted Mean gives different levels of importance (weights) to different values, instead of treating every value equally.

**Formula**: Weighted Mean = Σ(w × x) / Σw

where w = weight of each value, x = value

**Real-World Example**: A student's final grade is a weighted mean of assignment scores, midterm, and final exam, where the final exam carries more weight.

**Business Example**: A company calculates the weighted average price of raw materials purchased from different suppliers at different prices and quantities.

**AI/ML Example**: In ensemble models, predictions from multiple models are combined using a weighted mean, giving more weight to better-performing models.

In [4]:
scores = [70, 85, 90]     
weights = [0.2, 0.3, 0.5]  

weighted_mean = sum(s*w for s, w in zip(scores, weights))
print("Weighted Mean:", weighted_mean)

Weighted Mean: 84.5


### Interpretation:
Even though the simple average of 70, 85, and 90 would be 81.67, the weighted mean (84.5) is higher because the final exam (90), which carries the most weight (50%), pulls the result up.

## 05. Geometric Mean
**Definition**: Geometric Mean is the n-th root of the product of n values. It is used when values are multiplicative in nature, such as growth rates or ratios.

**Formula**: Geometric Mean = (x1 × x2 × ... × xn) ^ (1/n)

**Real-World Example**: Population growth rates over several years are averaged using the geometric mean, not the simple mean.

**Business Example**: A company calculates the geometric mean of its yearly revenue growth rates (e.g., 10%, 15%, 5%) to find the average growth rate over multiple years.

**AI/ML Example**: Geometric mean is used to evaluate model performance across multiple metrics (e.g., combining precision and recall) since it is less sensitive to extreme values than the simple mean.

In [5]:
%pip install scipy

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 26.1.2 -> 26.2
[notice] To update, run: C:\Users\hemak\AppData\Local\Programs\Python\Python314\python.exe -m pip install --upgrade pip


In [6]:
from scipy.stats import gmean

growth_rates = [1.10, 1.15, 1.05]  

geo_mean = gmean(growth_rates)
print("Geometric Mean:", geo_mean)
print("Average Growth Rate: {:.2f}%".format((geo_mean - 1) * 100))

Geometric Mean: 1.0992419018963262
Average Growth Rate: 9.92%


### Interpretation:
The average yearly growth rate over the three years is approximately 9.96%, which correctly accounts for compounding — unlike a simple mean, which would overstate the true average growth.

## 06. Harmonic Mean
**Definition**: Harmonic Mean is the reciprocal of the average of reciprocals of the values. It is used when dealing with rates, such as speed or ratios.

**Formula**: Harmonic Mean = n / (Σ(1/x))

where n = number of values, x = each value

**Real-World Example**: If a car travels the same distance at different speeds (e.g., going and returning), the harmonic mean gives the correct average speed for the whole trip.

**Business Example**: A company uses harmonic mean to calculate the average price-to-earnings (P/E) ratio across a portfolio of stocks.

**AI/ML Example**: The F1-Score in classification model evaluation is the harmonic mean of Precision and Recall, ensuring both are balanced fairly.

In [9]:
from scipy.stats import hmean

speeds = [60, 40] 

harmonic_mean = hmean(speeds)
print("Harmonic Mean (Average Speed):", harmonic_mean)


Harmonic Mean (Average Speed): 47.99999999999999


### Interpretation:
Even though the simple average of 60 and 40 would be 50, the correct average speed for the round trip is 48 km/h, because harmonic mean properly accounts for the fact that more time was spent traveling at the slower speed.

## Interview Questions and Answers

**1. What is the difference between Mean, Median, and Mode? When would you use each?**
Mean is the arithmetic average of all values, Median is the middle value when data is sorted, and Mode is the most frequently occurring value. Mean is best for symmetric data without outliers, Median is preferred for skewed data or data with outliers, and Mode is useful for categorical data or identifying the most common occurrence.

**2. Why is the median preferred over the mean when a dataset has outliers?**
The mean is calculated using every value, so extreme values (outliers) pull it in their direction, distorting the true "typical" value. The median only depends on the middle position(s) of sorted data, so it stays stable and gives a more representative central value even when outliers are present.

**3. Can a dataset have more than one mode? Give an example.**
Yes. A dataset can be bimodal (two modes) or multimodal (multiple modes). For example, in the dataset [2, 3, 3, 5, 5, 7], both 3 and 5 occur twice and are equally frequent, making this dataset bimodal.

**4. What is the difference between a simple mean and a weighted mean?**
A simple mean treats every value equally, while a weighted mean assigns different levels of importance (weights) to different values. Weighted mean is used when some values should contribute more to the final result, such as a final exam counting more than a quiz toward a course grade.

**5. When would you use Geometric Mean instead of Arithmetic Mean?**
Geometric mean is used when values are multiplicative in nature, such as growth rates, investment returns, or ratios, especially when the values compound over time. Arithmetic mean can overstate the true average in such cases because it doesn't account for compounding.

**6. Why is Harmonic Mean used for rates like speed, rather than Arithmetic Mean?**
Harmonic mean gives more weight to smaller values, which correctly reflects situations where more time or resources are spent at lower rates. For example, when averaging speeds over equal distances, harmonic mean accounts for the fact that traveling slower takes proportionally more time, giving a mathematically accurate average.

**7. What is the relationship between Arithmetic Mean, Geometric Mean, and Harmonic Mean?**
For any dataset of positive numbers, AM ≥ GM ≥ HM, with equality holding only when all values in the dataset are identical. This inequality is a fundamental result in mathematics and statistics.

**8. Why is mode considered the only measure of central tendency applicable to categorical/nominal data?**
Mean and median require numerical, ordered data to calculate meaningfully, but mode simply counts frequency, which works even for non-numeric categories like colors, city names, or product types.

**9. In machine learning, how are Mean, Median, and Mode used for handling missing values?**
Mean is used to fill missing numerical values when data is roughly symmetric and without major outliers. Median is used instead when the numerical data is skewed or has outliers. Mode is used to fill missing categorical values with the most frequently occurring category.

**10. Give a real-world example where the mean can be misleading.**
Average (mean) household income in a region can be misleading if a few extremely wealthy households are included, since they inflate the mean far above what a typical household actually earns. The median household income is usually reported instead for a fairer picture.

## Self Reflection

- **Mean** is intuitive and widely used, but I need to be cautious when data has extreme values, since it can misrepresent the typical case.
- **Median** is a more reliable measure for real-world data like incomes or prices, where outliers are common.
- **Mode** is often overlooked but essential for categorical features and understanding frequency patterns.
- **Weighted Mean** taught me that not all data points deserve equal importance — context (like exam weightage) should shape the calculation.
- **Geometric Mean and Harmonic Mean** were the most eye-opening, since I previously assumed the simple mean was always correct. I now understand that growth rates and rate-based data (like speed) require specialized averages to get mathematically accurate results.